woocommerce/woocommerce-gateway-stripe
# Implement re-connect flow using the Mkt App Oauth as provider

Fixes #3338

## Changes proposed in this Pull Request:

This PR adds support for the Stripe App OAuth flow for account re-authentication.

If an account is already connected to a different platform account, it can not complete the normal Connect OAuth flow, the account can't be selected in the Stripe Account selection page... and it shows the following message: `This account is connected to another platform. Please create a new account.`. 
<img width="469" alt="Screenshot 2024-08-08 at 3 19 09 AM" src="https://github.com/user-attachments/assets/15c6ddc8-f168-48c5-a601-7c4dcb75147c">

For merchants in this situation, we use the Stripe App OAuth flow for authentication:
![Screenshot 2024-08-06 at 11 08 04 PM](https://github.com/user-attachments/assets/e965eb67-36fa-4bf6-a523-de178fb8707c)


## Testing instructions

1. Navigate to WP-Admin > WooCommerce > Settings > Payments > Stripe
    > /wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe
2. Connect a test account (if not connected already)
3. Click on `Configure connection` and mouse over the `Create or connect a test account button` (or click it), check that the URL is `https://connect.stripe.com`
    <img width="823" alt="Screenshot_2024-08-05_at_2_15_19 AM" src="https://github.com/user-attachments/assets/17dfd03c-9b27-4572-9d77-d14b6d148fcb">
4. To test the Stripe App OAuth flow, we need to configure the plugin to use keys from an account connected to another platform. If you don't have an account connected to a 3rd party platform, ping me in slack.
5. Reload the settings page, and check that the `accountId` was updated (1), and mouse over the `Create or connect a test account` button, check that the URL is `https://marketplace.stripe.com`:
    <img width="1419" alt="Screenshot_2024-08-05_at_2_26_59 AM" src="https://github.com/user-attachments/assets/edbd3f52-41a7-41ad-9971-7ee887b4b114">
6. Click the `Create or connect a test account` button.
7. Login to Stripe (if needed)
8. Complete the Stripe App Oauth process, selecting an account, accepting the permissions and installing the app:
    ![image (2)](https://github.com/user-attachments/assets/9b754244-4fb2-4ff5-b4ff-535ea41e79bf)
9. Check that the flow was successful by verifying the keys in the plugin settings:
    - Check that the `test_secret_key` starts with **`rk_`**
        <img width="1261" alt="354987227-859673ca-7f1b-4b9c-b779-5fc08425ac96-2" src="https://github.com/user-attachments/assets/bf41a7c3-8b1f-4548-8a0b-d595fa4e877a">
    - Check the `test_connection_type` was saved in the settings, with the value `app`:
        <img width="742" alt="354987278-a1ca0376-f8dd-4db4-800d-a1e875e1c7ed-2" src="https://github.com/user-attachments/assets/c900211a-9060-4661-a4a1-d55ead133927">

_NOTE_: The account selected in the Stripe App OAuth does not need to be the same one it was configured before... an account connected to our platform account can be selected; in this case, we'll still get the restricted keys and save the origin of the keys as `app`, but when generating the links for re-authentication (in the configure connection modal) we'll show the connect links as we check the controller to decide which flow to use.

---


**Post merge**
